chore: refactor files and added eslint

zaida04 opened the pull request:
Hey Sefath, noticed you made this on your insta, so I thought i'd take a look and thought perhaps I could help a bit since I do a bit of projects with ExpressJS and Mongoose. I made a bit of changes detailed below. Feel free to open comments and such. And check out my profile if you want

- Added Linting, using eslint.
- Added .eslintrc.json with linting rules (Inspected files to find rules)
- Added npm start, npm test, and npm run lint in package.json
- Added dependency in package.json, eslint
- Made app.js more consistent, and most routes now use async so that it's non-blocking (more details at the bottom)
- Made promisifedFunctions.js, will populate with other async functions that take callback and can be turned into promises. (so far added jwt.verify)
- Not quite sure why a local path was being required in mongoose.js, but I assumed it was to use dotenv, so I simplified it a bit

still doing some stuff but i thought i'd open the pr just cause. 
**Don't merge yet, I'm still adding commits cause it's morning and i'm tired. Will add comment when ready for merge.**

Commit:
A whole lotta

- Added Linting, using eslint.
- Added .eslintrc.json with linting rules (Inspected files to find rules)
- Added npm start, npm test, and npm run lint in package.json
- Added dependency in package.json, eslint
- Made app.js more consistent, and most routes now use async so that it's non-blocking (Will detail change in PR)
- Made promisifedFunctions.js, will populate with other async functions that take callback and can be turned into promises. (so far added jwt.verify)
- Not quite sure why a local path was being required in mongoose.js, but I assumed it was to use dotenv, so I simplified it a bit

zaida04 commented on api/app.js:
```diff
@@ -1,21 +1,26 @@
 //Cross-Origin Resource Sharing Middleware
 //https://enable-cors.org/server.html
-app.use(function(req, res, next) {
+app.use((req, res, next) => {
```
Made using es6 arrow functions. Assuming you won't need to use the `this` keyword. Just thought it would make it more consistent

zaida04 commented on api/app.js:
```diff
@@ -1,21 +1,26 @@
 /* ---Load Mongoose Models--*/
-const { List, Task, User } = require('./db/models');
-const jwt = require('jsonwebtoken');
+const {
```
Sorry, VSCode auto formats onto newlines, tried to stop it but oh well

zaida04 commented on api/app.js:
```diff
@@ -1,21 +1,26 @@
+    User
+} = require('./db/models');
+const {
+    verify
```
Instead of having functions that take a callback function, I promisifed the function to avoid callback hell.

zaida04 commented on api/app.js:
```diff
@@ -1,21 +1,26 @@
-
+require('dotenv').config();
 
+const port = process.env.PORT || 3000;
```
If you do make other commits and merge this into deployment, using this will allow it to run on heroku, seeing as heroku provides a port and gives that info through an env variable, if not provided then just fallback to port 3000

zaida04 replied:
**REVERTED**

zaida04 commented on api/app.js:
```diff
@@ -30,38 +35,37 @@ app.use(function(req, res, next) {
 
 // Verify Refresh Token Middleware (which will be verifying the session)
-let verifySession = (req, res, next) => {
+let verifySession = async (req, res, next) => {
```
made async function

zaida04 commented on api/app.js:
```diff
@@ -30,38 +35,37 @@ app.use(function(req, res, next) {
-    User.findByIdAndToken(_id, refreshToken).then((user) => {
-        console.log(user);
+    try {
+        let user = await User.findByIdAndToken(_id, refreshToken)
```
converted from using .then to using async/await

zaida04 commented on api/app.js:
```diff
@@ -30,38 +35,37 @@ app.use(function(req, res, next) {
-            return Promise.reject({
-                'error': 'User not found. Make sure that the refresh token and user id are correct'
-            });
+            throw new Error('User not found. Make sure that the refresh token and user id are correct')
```
Since the function is now async, and async functions return promises, throwing an error is the exact same as rejecting the promise

zaida04 commented on api/app.js:
```diff
@@ -76,10 +80,7 @@ let verifySession = (req, res, next) => {
-                    // refresh token has not expired
-                    isSessionValid = true;
-                }
+                isSessionValid = Boolean(User.hasRefreshTokenExpired(session.expiresAt))
```
The if wasn't neccesary, since we can just cast the return to a Boolean based on if it returns a truthy or falsy value

zaida04 commented on api/app.js:
```diff
@@ -109,60 +107,67 @@ let verifySession = (req, res, next) => {
-    }).then((lists) => {
-        res.send(lists);
-    }).catch((err) => {
+    try {
```
.then => async/await

zaida04 commented on api/app.js:
```diff
@@ -109,60 +107,67 @@ let verifySession = (req, res, next) => {
-    newList.save().then((listDoc) => {
-        //full list document is returned
-        res.send(listDoc);
+        "title": req.body.title,
```
Wasn't really a need to initialize a variable, and i quoted the props cause of eslint

zaida04 commented on api/app.js:
```diff
@@ -109,60 +107,67 @@ let verifySession = (req, res, next) => {
  */
-app.post('/lists', authenticate, (req, res) => {
-    let title = req.body.title;
+app.post('/lists', authenticate, async (req, res) => {
```
.then => async/await

zaida04 commented on api/app.js:
```diff
@@ -109,60 +107,67 @@ let verifySession = (req, res, next) => {
+        "title": req.body.title,
+        "_userId": req.user_id
     });
+    res.send(await newList.save());
```
more compact

zaida04 commented on api/app.js:
```diff
@@ -109,60 +107,67 @@ let verifySession = (req, res, next) => {
-    }).then(() => {
-        res.send({ message: 'updated successfully' });
-    });
+    try {
```
.then => async/await

zaida04 commented on api/functions/promisifiedFunctions.js:
```diff
@@ -0,0 +1,6 @@
+const jwt = require('jsonwebtoken');
```
You can add more packages here that do involve functions that take a callback, and add them to the module.exports

zaida04 commented on api/db/mongoose.js:
```diff
@@ -1,7 +1,7 @@
 
 // This file will handle connection logic to MongoDB Database
-const mongoose = require ('mongoose');
+const mongoose = require('mongoose');
```
linter did this

zaida04 commented on api/db/mongoose.js:
```diff
@@ -1,7 +1,7 @@
-require('../node_modules/dotenv/config');
+require('dotenv').config();
```
local path => npm recognized path

zaida04 commented on api/app.js:
```diff
@@ -368,21 +345,22 @@ app.post('/users/login', (req, res) => {
 
-app.listen(3000,  () => {
-    console.log('server is listening on port 3000');
+app.listen(port, () => {
```
use port variable instead of hard coded number

zaida04 commented on api/app.js:
```diff
@@ -368,21 +345,22 @@ app.post('/users/login', (req, res) => {
-app.listen(3000,  () => {
-    console.log('server is listening on port 3000');
+app.listen(port, () => {
+    console.log(`server is listening on port ${port}`);
```
String concatenation => Template Literals

zaida04 commented on api/app.js:
```diff
@@ -368,21 +345,22 @@ app.post('/users/login', (req, res) => {
 /* HELPER METHODS */
-let deleteTasksFromList = (_listId) => {
-    Task.deleteMany({
+let deleteTasksFromList = async (_listId) => {
```
async function

zaida04 commented on api/app.js:
```diff
@@ -368,21 +345,22 @@ app.post('/users/login', (req, res) => {
-let deleteTasksFromList = (_listId) => {
-    Task.deleteMany({
+let deleteTasksFromList = async (_listId) => {
+    await Task.deleteMany({
```
.then => async/await

zaida04 commented on api/app.js:
```diff
@@ -109,60 +107,67 @@ let verifySession = (req, res, next) => {
-        });
+ */
+app.delete('/lists/:id', authenticate, async (req, res) => {
+    let removedListDoc = await List.findOneAndRemove({
```
.then => async/await

zaida04 commented on api/app.js:
```diff
@@ -174,126 +179,93 @@ app.delete('/lists/:id', authenticate, (req, res) => {
-            res.send(tasks);
-        });
+app.get('/lists/:listId/tasks', authenticate, async (req, res) => {
+    res.send(await Task.find({
```
more compact and no .then

zaida04 commented on api/app.js:
```diff
@@ -174,126 +179,93 @@ app.delete('/lists/:id', authenticate, (req, res) => {
-    });
-})
+app.get('/lists/:listId/tasks/:taskId', authenticate, async (req, res) => {
+    res.send(await Task.findOne({
```
more compact and no .then

zaida04 commented on api/app.js:
```diff
@@ -174,126 +179,93 @@ app.delete('/lists/:id', authenticate, (req, res) => {
-        }
-    }).then((canCreatetask) => {
-        if(canCreatetask) {
+    let list = await List.findOne({
```
.then => async/await

zaida04 commented on api/app.js:
```diff
@@ -174,126 +179,93 @@ app.delete('/lists/:id', authenticate, (req, res) => {
+        "_id": req.params.listId,
+        "_userId": req.user_id
     })
+    if (list) {
```
more compact, and no .then

zaida04 commented:
^ I commented reason for most of my changes, still not done yet tho

Commit:
Added .env to gitignore (don't accidentally wanna commit that)
Added lint github workflow (for prs and such)
quoted prop in mongoose.js lol

Commit:
took more of a look and realized it wasn't a good idea to have port rely on env variable

Shefuchow commented:
@zaida04 Dude you're a beast haha, thanks for looking through this! i appreciate it, i'll take a look at everything after work :)

ps. hope you and you're fam are doing well, take care brother

Commit:
Fully converted all functions to async

zaida04 commented:
Finished the conversion of all functions to async functions, and converted all uses of .then and .catch to use async/await with try/catch blocks. A lot more compact now, and I'm too lazy comment every single change again, so I'll just let the version history show.

zaida04 commented:
Once I test it, it'll be ready for merge

Commit:
Did some more stuff cause I had time

zaida04 commented:
Moved linting to root directory, for easier access. 
Made package.json in root, for using concurrently

**Two new commands, to make stuff easier, need to do `npm i` in root directory for it to work.**
- `npm run start` (will start both the api, and the frontend, in one command)
- `npm run custom-install` (will run npm install in both the api and frontend directory, in one command)

Moved linting to root directory too, `npm lint`

Commit:
change for postinstall

zaida04 commented:
Small change, replace custom-install with postinstall, which automatically runs after you do npm install in the root of the project. Easy way to just have to do npm install once for the entire project.

Commit:
add eslint to package.json

Commit:
small fixes

zaida04 commented:
bump for this

zaida04 commented:
Hey, I'm sure you forget about this a while ago, but the pr is still open if you wanna review and merge

zaida04 commented:
Closing cause stale. You can reopen and merge later.
